### ISSUE TITLE
convert torch model to mxnet model

### ISSUE THREAD

**Issue by loweew** [dmlc/mxnet#1567]
I have a torch model from here: https://github.com/facebook/fb.resnet.torch which I fine-tuned on my own data set. I have several mxnet models from inception-v3 imagenet which I've also fine-tuned using various optimizers. I'd like to ensemble these models with several torch models, but would like to continue using mxnet. Is there a way to either read in the torch models using mxnet or convert the torch model t7 files to mxnet params files? I've already implemented a flask server for mxnet predictions which I'd like to put into production soon, so staying with mxnet is greatly desired.

**Comment by piiswrong** [comment]
Thanks for your interest.
If you can read t7 files from python, it should be very easy to create a mxnet param file with mx.nd.save. It's basically a dictionary of param name to ndarrays.
One quick hack would be to write a lua script that save model to ascii format and read from python.

BTW, I think a lot of people would like to use a flask prediction server. Could you release your code?

**Comment by xgastaldi** [comment]
If you don't mind finetuning again, another solution could be to start from the pre-trained caffe model that Kaiming released (https://github.com/KaimingHe/deep-residual-networks) and then import the fine-tuned model in mxnet with the caffe model converter

**Comment by loweew** [comment]
I would like to be able to fine-tune with mxnet to keep things consistent, but the convert_model.py fails with errors already mentioned in #1419 

Are there any plans to update the conversion scripts? 

Regarding the flask server, I'll be happy to release that as soon as I'm convinced it's doing what I want.

**Comment by futurely** [comment]
Steps to directly fine tune an MXNet model: https://github.com/dmlc/mxnet/issues/479#issuecomment-153809302.

**Comment by loweew** [comment]
Yes, the actual fine-tuning is trivial and I've successfully done that on several pre-trained mxnet models. The current issue is just that the conversion scripts for caffe models are not up-to-date. So, using the caffe models as suggest by @xgastaldi isn't possible at the moment using the caffe model converter in its current state. I'll have some time to look into this during the weekend. If I figure out how to fix the scripts to recognize the BatchNorm layer appropriately, I'll open a PR.

**Comment by futurely** [comment]
It is in fact quite non-trivial to fine tune a pre-trained model whether the model is trained with MXNet or converted from other frameworks. An example is https://github.com/dmlc/mxnet/pull/1568.

MXNet has a powerful execution engine. But there's a large room to improve the usability of its API. Fine tuning is an example of MXNet's ["underlying architecture bleeding through into the semantic model"](https://cloud.google.com/dataflow/blog/dataflow-beam-and-spark-comparison). It requires significant efforts to redesign the API to stop the bleeding.

**Comment by taoari** [comment]
I have re-designed the finetune api, which is available at https://github.com/taoari/mxnet/blob/finetune-api/example/finetune-api/README.md, I will send another PR once PR1570 and PR1617 are merged.

It is caffe-flavored, and is quite easy to use. For example, fine-tuning Caltech101 with the inception-bn model:

```
train_imagenet.py --solver finetune_caltech101_inception.yml --gpus 0
```
* finetune_caltech101_inception.yml

```
network: inception-bn

data_dir: ../data/caltech101
dataset: caltech101
num_classes: 102
num_examples: 3060
batch_size: 25

lr: 0.0005
lr_factor: 0.1
lr_factor_epoch: 320
momentum: 0.9
wd: 0.0005

num_epochs: 800
kv_store: local

model_prefix: finetune_caltech101_inception
finetune_from: ../share/inception-bn/Inception_BN-0039.params
checkpoint_epoch: 80
eval_epoch: 16
```

**Comment by futurely** [comment]
Another ad-hoc method is two-stage training. To only train newly added layers such as a fully connected layer with a number of classes different from that of the original network, the first stage trains a cloned network with BlockGrad added after the internal layers. The internal layers parameters are still updated for the reason in https://github.com/dmlc/mxnet/issues/1340#issuecomment-174166248. The second stage trains the cloned network with all the BlockGrad symbols removed.

Inspired by the following codes.
https://github.com/dmlc/mxnet/blob/d717c42025eeb0259e1e3e63d47a84f67a3740d3/tools/accnn/utils.py
https://github.com/dmlc/mxnet/blob/cb8300739de673ed81540192f5129a58ff5ba456/tools/accnn/acc_conv.py
https://github.com/dmlc/mxnet/issues/1538#issuecomment-196691988

**Comment by taoari** [comment]
@futurely My experiments show that two-stage training is not a good idea. If you first train a model only allowing changes of the final fully connected layers, it is OK. But in the second stage, the model will explode quickly after several iterations.

**Comment by loweew** [comment]
My point was that I'm not having any trouble with fine tuning networks. I've been able to accomplish this previously having already referenced the issues you are linking from your previous helpful posts. The manner in which fine tuning is performed was not why this issue was opened. I don't disagree that the fine tuning API could use some refactoring. But, again, this is not the point of this issue. I'm closing now that focus had been derailed.

**Comment by futurely** [comment]
https://github.com/dmlc/mxnet/pull/1647

**Comment by savy-91** [comment]
Hi @loweew , did you manage to convert a torch model to mxnet in the end?
